lib/open_project/scm/adapters/subversion.rb

```diff
@@ -0,0 +1,289 @@
+              next unless line =~ %r{^\s*(\d+)\s*(\S+)\s(.*)$}
+              blame.add_line($3.rstrip, Revision.new(identifier: $1.to_i, author: $2.strip))
+            end
+          end
+          blame
+        end
+
+        private
+
+        ##
+        # Builds the SVM command arguments around the given parameters
+        # Appends to the parameter:
+        # --username, --password    if specified for this repository
+        # --no-auth-cache           force re-authentication
+        # --non-interactive         avoid prompts
+        def build_svn_cmd(args)
+          if @login.present?
+            args.push('--username', shell_quote(@login))
+            args.push('--password', shell_quote(@password)) if @password.present?
+          end
+
+          args.push('--no-auth-cache', '--non-interactive')
+        end
+
+        def xml_capture(cmd, opts = {})
+          output = capture_svn(cmd, opts)
+          doc = Nokogiri::XML(output)
+
+          # Yield helper methods instead of doc
+          yield doc
+        end
+
+        def extract_entry(entry, path)
+          revision = extract_revision(entry.at_xpath('commit'))
+          kind, size, name = parse_entry(entry)
+
+          # Skip directory if there is no commit date (usually that
+          # means that we don't have read access to it)
+          return if kind == 'dir' && revision.time.nil?
+
+          Entry.new(
+            name: URI.unescape(name),
+            path: ((path.empty? ? '' : "#{path}/") + name),
+            kind: kind,
+            size: size.nil? ? nil : size.to_i,
+            lastrev: revision
+          )
+        end
+
+        def parse_entry(entry)
+          kind = entry['kind']
+          size = entry.xpath('size').text
+          name = entry.xpath('name').text
+
+          [kind, size, name]
+        end
+
+        def build_path(entry)
+          {
+            action: entry['action'],
+            path: entry.text,
+            from_path: entry['copyfrom-path'],
+            from_revision: entry['copyfrom-rev']
+          }
+        end
+
+        def extract_revision(commit_node)
+          # We may be unauthorized to read the commit date
+          date =
+            begin
+              Time.parse(commit_node.xpath('date').text).localtime
+            rescue ArgumentError
+              nil
+            end
+
+          Revision.new(
+            identifier: commit_node['revision'],
+            time: date,
+            message: commit_node.xpath('msg').text,
+            author: commit_node.xpath('author').text
+          )
+        end
+
+        def fetch_revision_entries(identifier_from, identifier_to, options, path, &block)
+          path ||= ''
+          identifier_from = numeric_identifier(identifier_from, 'HEAD')
+          identifier_to = numeric_identifier(identifier_to, 1)
+          cmd = ['log', '--xml', '-r', "#{identifier_from}:#{identifier_to}"]
+          cmd << '--verbose' if options[:with_paths]
```

Comment:
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.